### DISCUSSION TITLE
MemoryTransport vs NetworkTransport

### DISCUSSION THREAD

**Discussion by Tanguille** [libp2p/rust-libp2p#4465] [Q&A] [answered]
I have read the general `libp2p` docs. I am now trying to implement PubSub and came across a MemoryTransport. When I did some further digging I found you have Network- and MemoryTransport. Curiously I didn't read anything about that in the docs so I was wondering if someone could explain or show me where I can find the resources to learn about this. 

Implementing `libp2p` with rust is going rough at the moment. So any tutorials, video's, more concise examples are always welcome aswell!

**Comment by dariusc93** [comment]
`MemoryTransport` is used for communicating between two or more nodes spawned within the same process. Useful for writing test without needing to utilize tcp or udp. I will admit the docs itself is a bit lacking on it from what I recall (and could be improved among other things), but the tests related to it is pretty self-explanatory.

**Reply by Tanguille** [reply]
So MemoryTransport is just for development purposes and has no use in actual deployment?

**Reply by kehiy** [reply, marked as answer]
The `MemoryTransport` is mainly used for testing and debugging purposes.

**Comment by kehiy** [comment]
hi @Tanguille,

The main difference between them is that the `MemoryTransport` is used for communication between the same process, but the `NetworkTransport` is used for communication between different processes over a network.

About learning and resource, the best way is docs and after it go through reading code examples!
It's how you can do it. (don't forgot to write codes after)

**Reply by Tanguille** [reply]
I am reading through the docs but the explanation is often a bit limited to understand what I need to do. For example after reading [this](https://docs.rs/libp2p-dns/latest/libp2p_dns/) I still have no idea why you would need dns functionality. Is it for supporting nodes that run from a reverse proxy or something? It isn't clearly mentioned  I feel like. 

I have been using https://github.com/libp2p/rust-libp2p/tree/master/misc/server/src and https://github.com/mxinden/libp2p-lookup/ as examples but they are a bit limited to be able to see how all features in libp2p actually work. Are there any other sources you could recommend to learn how to setup a proper p2p network in rust?

**Reply by kehiy** [reply]
I'm not sure about `libp2p-dns`, but in general my approach to lean this stuff is first read the essentials, then examples and start code something easy.
after this I go thorough other stuff step by step. 
by this way I can make clear all this stuff in my mind.

you can try this too I think!

about examples, you can see [iroh](https://github.com/n0-computer/iroh) as a ipfs implementation in rust.

**Reply by Tanguille** [reply]
Yeah that's exactly what I'm trying to do but it seems that doing something easy in libp2p is very hard since you need a lot of parts to test if what you're trying is actually working. 

That iroh project looks very interesting, thanks!
